Add MODE command & various text mode improvements

johnnovak opened the pull request:
# Description

Implements https://github.com/dosbox-staging/dosbox-staging/issues/3670.

I ended up on a wild goose chase with this one again... Took a grand tour in the bowels of the VGA code one more time, and I've done some cleanups & minor refactors along the way. After a few rewrites, the final solution ended up being quite simple, but I think it's worthwhile to keep the cleanups anyway. The real changes are in the last few commits.

I ended up adding custom BIOS modes for the various non-standard text modes the MODE command supports (just on VGA, though). This has several benefits:

- I could fine-tune the timings to make these modes fill the screen completely.
- It turns out this is the most compatible way, as the "traditional way" of setting up a standard BIOS mode first and then tweaking it further results in some weirdness (e.g., the tweaks can, under some circumstances, "bleed into" subsequent standard INT 10h mode changes. This might be an emulation bug, but I'm really hesitant to change the VGA text mode stuff significantly).
- You can use these modes in programs that let you set custom BIOS modes. DOS Navigator 1.51 is a prime example—you can configure DN to set any valid BIOS text mode on startup _and_ restore it whenever an executable launched from DN exits. Like a pro 😎

These custom modes are safe additions as programs can't list the available BIOS modes pre-VESA, so they wouldn't even know they exist.

Interestingly, 132-column modes on the Tseng and Paradise adapters seem to be at least partially broken—they can be set all right, but the DOS prompt will behave rather erratically (e.g., disappearing cursor, black screen, etc.) I verified this going back to 0.79.1; it's not a recent regression, and even tested it in SVN. Yep, it's broken in SVN too, so it wasn't our doing.  So I've simply excluded the 132-column modes from the Tseng and Paradise.

I've also tweaked the timings of some of the existing modes a little bit so the image fills the screen better, and made the 132-column modes uniformly "widescreen" (that's a nice use-case of the 132-column modes on modern screens).

Also fixed a bug where the 8x14 font was set instead of the 9x14 variant on VGA for BIOS modes that use 14-line fonts. According to the "Programmer's Guide to the EGA, VGA and Super VGA Cards (3rd ed.)", VGA cards prefer the 9-dot fonts in all modes if they're available. Well, I guess I could test it on my S3 cards, but I'm too lazy to do that 😄 In any case, 9x14 looks better, and I'm 90% sure this is authentic.

8-dot fonts can still be forced on VGA with the `vga_8dot_font` setting (@FeralChild64 can be happy as he seemed to like 8-dot fonts). I tried adding support for selecting 8 vs 9-dot fonts, but it turned out to be a bad idea as it's hard to do correctly, so I've dropped the feature.

Help text after setting `mode 80x30`:

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/aca35e36-0a15-46d6-9527-29e2558b61fb">


# Manual testing

Tested:

- All modes on all supported graphics adapters, including `vga_8dot_font = on` on VGA adapters
- All error scenarios on different graphics adapters
- The typematic rate settings

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

rderooy commented on src/dos/program_mode.cpp:
```diff
@@ -0,0 +1,389 @@
+};
+// clang-format on
+
+static std::set<std::string> valid_video_modes_hercules = {"40x25"};
```
80x25 instead?

johnnovak replied:
Good pickup, yes 😄

MasterO2 commented:
One small suggestion, @johnnovak:  Specify in the `MODE` explanation page that the default display mode is 80x43.

johnnovak commented:
> One small suggestion, @johnnovak: Specify in the `MODE` explanation page that the default display mode is 80x43.

80x25 is the default mode. I'll consider mentioning it, but probably I won't.

FeralChild64 commented on src/dos/program_mode.h:
```diff
@@ -0,0 +1,50 @@
+	{
+		AddMessages();
+		help_detail = {HELP_Filter::All,
+		               HELP_Category::File,
```
`HELP_Category::File` -> `HELP_Category::Misc`

You can see the result in `help /all` command output, the parameter decides upon section the command is listed in.

FeralChild64 commented on src/dos/program_mode.cpp:
```diff
@@ -0,0 +1,451 @@
+	// and 1056 and 1188 in 132-column modes.
+	clocking.is_eight_dot_mode = (width == FontWidth::_8dot);
+
+	IO_Write(0x3c4, ClockingModeRegisterIndex);
```
We have constants for I/O addresses in `inout.h`; it would be nice to continue using this convention.

FeralChild64 commented:
Just two remarks. I've played a little bit with the command - it mostly works, but a lot of DOS software cannot handle the extended text modes, sadly. There are, however, problems with `8dot` and `9dot` commands.
1. If `vga_8dot_font` is enabled in config file, switching between 8/9 dot modes does not work.
2. 8 dot font set by the `vga_8dot_font` option is different than the one set by the `8dot` parameter - see the screenshots below, look at _mo_ letters in _mode_ word. IMHO it's the `vga_8dot_font` option that did it right.

![1](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/170c0e84-6560-4450-a8b6-f45ebea3744a)
![2](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/68df78aa-6907-4d6b-82ef-59a311783f3a)

johnnovak commented:
> Just two remarks. I've played a little bit with the command - it mostly works, but a lot of DOS software cannot handle the extended text modes, sadly.

Yeah, that's expected. And practically all graphical programs reset the 80x25 mode on exit anyway. Maybe I could hook into the exit process to set my own mode, but not sure it's worth it.

>     1. If `vga_8dot_font` is enabled in config file, switching between 8/9 dot modes does not work.

That's expected. That flag emulates old VGA boards that did not have the 9dot fonts. Effectivy it forces 8dot mode. But I could do better and raise a warning or error if it's set.

>     2. 8 dot font set by the `vga_8dot_font` option is different than the one set by the `8dot` parameter - see the screenshots below, look at _mo_ letters in _mode_ word. IMHO it's the `vga_8dot_font` option that did it right.

I will look into this, I haven't noticed 😄

FeralChild64 commented:
I have also noticed that the text cursor blinks much more slowly in modes like 132x30 than in 80x25.

johnnovak commented:
> I have also noticed that the text cursor blinks much more slowly in modes like 132x30 than in 80x25.

Indeed, you have a good eye. I've never noticed.

That's because the cursor blink rate is apparently tied to the refresh rate of the video mode.

80x30 -> 60 Hz
80x25 -> 70 Hz
80x43 -> 81 Hz

132x25 -> 37 Hz
etc...

The non-VESA 132 Hz modes have sub-40 Hz refresh rates. I suspect this is most likely an emulation inaccuracy as most non-multisync VGA monitors couldn't sync below 50 Hz. In any case, I'm not gonna delve deeper... 😅 

What I could do, though, is to use the VESA 132-column modes instead as those are uniformly around 69-70 Hz with the more flexible VESA timings. Then I could convert my extra 132-column modes to VESA modes.

johnnovak commented:
> Just a curiosity of mine, not related to the code here - to me the DOS prompt cursor (the dash that blinks) doesn't look like any of the ASCII characters. Am I wrong - is it an ASCII character? Or is it custom-drawn somehow?

Google is a wonderful thing 😎  This is what my guess was before looking it up, btw 😏 

https://retrocomputing.stackexchange.com/a/5604

japsmits commented on src/ints/int10_modes.cpp:
```diff
@@ -1949,51 +2055,67 @@ bool INT10_SetVideoMode(uint16_t mode)
+		IO_Write(crtc_base,0x3a);
+		IO_Write(crtc_base+1,reg_3a);
+
+		IO_Write(crtc_base,0x31);
+		// Enable banked memory and 256k+ access
+		IO_Write(crtc_base+1,reg_31);
```
Where does the comment belong? Should it be something like this:
```suggestion
		// Enable banked memory and 256k+ access
		IO_Write(crtc_base,0x31);
		IO_Write(crtc_base+1,reg_31);
```
If so, this comment goes for quite a few other places as well.

johnnovak replied:
I'm not gonna spend more time on these reactors now, just making small gradual improvements for now.

japsmits replied:
Understandable, but then it might be preferable to revert these changes completely. The placing of the comments does not match the intention of the previous code, and if this is committed as is, the association between code and comment is lost.

johnnovak replied:
Yeah, I agree with fixing the comments. I thought you were referring to other cosmetic changes. Thanks for catching these.

japsmits commented on src/ints/int10_modes.cpp:
```diff
@@ -1647,84 +1706,112 @@ bool INT10_SetVideoMode(uint16_t mode)
+
+		// Color Graphics
+		att_data[0x10] = 0x01;
+		for (uint8_t ct = 0; ct < 16; ct++) {
```
Strictly speaking, it would have to become the following:
```suggestion
		for (uint8_t ct = 0; ct < 16; ++ct) {
```
But since this is obviously a reformat, this can wait for some other time. 😄

japsmits commented on src/ints/int10_memory.cpp:
```diff
@@ -22,112 +22,179 @@
+	IO_Write(0x3c4, 0x02);
+
+	// select plane 2 for writing
+	IO_Write(0x3c5, 0x04);
```
Formatting here is borked 🐔 . The comments are misplaced. I would expect something like:
```
// Put video adapter in planar mode

// select plane 2 for writing
IO_Write(0x3c4, 0x02);
IO_Write(0x3c5, 0x04);

// odd/even off in SEQ
IO_Write(0x3c4, 0x04);
IO_Write(0x3c5, 0x07);

...
```

japsmits commented on src/ints/int10_memory.cpp:
```diff
@@ -22,112 +22,179 @@
-static uint16_t map_offset[8]={
-	0x0000,0x4000,0x8000,0xc000,
-	0x2000,0x6000,0xa000,0xe000
+static std::vector<uint16_t> map_offset = {
+	0x0000, 0x4000, 0x8000, 0xc000,
+	0x2000, 0x6000, 0xa000, 0xe000
```
`static const std::array<uint16_t, 8>` instead of a vector? Or is this supposed to be expandable?

johnnovak replied:
Adding `const`, though.

japsmits commented on src/ints/int10_memory.cpp:
```diff
@@ -22,112 +22,179 @@
- /* e */ 0x00,  // Change to add new functions
- /* f */ 0x00   // reserved
+// clang-format off
+static uint8_t static_functionality[0x10] = {
```
```suggestion
static const std::array<uint8_t, 0x10> static_functionality = {
```

johnnovak replied:
Doing the `static const` part, but leaving it as a vector. I dislike having to specify the size; that's error-prone. `std::vector` is just an array if you never resize it, so personally I'd just default to `std::vector` everywhere.

japsmits commented on src/ints/int10.cpp:
```diff
@@ -632,34 +797,45 @@ static Bitu INT10_Handler(void) {
 	case 0xff:
-		if (!warned_ff) LOG(LOG_INT10,LOG_NORMAL)("INT10:FF:Weird NC call");
-		warned_ff=true;
+		if (!warned_ff) {
+			LOG(LOG_INT10, LOG_NORMAL)("INT10: FF:Weird NC call");
+		}
+		warned_ff = true;
```
```suggestion
		if (!warned_ff) {
			LOG(LOG_INT10, LOG_NORMAL)("INT10: FF:Weird NC call");
			warned_ff = true;
		}
```

johnnovak replied:
Okay, but bear in mind I'm just doing minimal cleanups/reformats as parts of this change. I intentionally did _not_ want to even begin to understand that code too much; this was just cosmetic cleanup.

japsmits commented on src/ints/int10_memory.cpp:
```diff
@@ -22,112 +22,179 @@
-	IO_Write(0x3ce,0x04);IO_Write(0x3cf,0x00); // select plane 0 for reading
-	IO_Write(0x3ce,0x05);IO_Write(0x3cf,0x10); // write mode 0, odd/even on in GFX
-	IO_Write(0x3ce,0x06);IO_Write(0x3cf,mono?0x0a:0x0e); // Bx000-BxFFF, odd/even on
+	// Return to normal text mode
+	IO_Write(0x3c4, 0x02);
+
+	// select planes 0&1 for writing
+	IO_Write(0x3c5, 0x03);
+	IO_Write(0x3c4, 0x04);
+
+	// odd/even on in SEQ
+	IO_Write(0x3c5, 0x03);
+	IO_Write(0x3ce, 0x04);
+
+	// select plane 0 for reading
+	IO_Write(0x3cf, 0x00);
+	IO_Write(0x3ce, 0x05);
+
+	// write mode 0, odd/even on in GFX
+	IO_Write(0x3cf, 0x10);
+	IO_Write(0x3ce, 0x06);
+
+	// Bx000-BxFFF, odd/even on
+	IO_Write(0x3cf, mono ? 0x0a : 0x0e);
```
👨‍🍳🐔 
```suggestion
	// Return to normal text mode

	// select planes 0&1 for writing
	IO_Write(0x3c4, 0x02);
	IO_Write(0x3c5, 0x03);

	// odd/even on in SEQ
	IO_Write(0x3c4, 0x04);
	IO_Write(0x3c5, 0x03);

	// select plane 0 for reading
	IO_Write(0x3ce, 0x04);
	IO_Write(0x3cf, 0x00);

	// write mode 0, odd/even on in GFX
	IO_Write(0x3ce, 0x05);
	IO_Write(0x3cf, 0x10);

	// Bx000-BxFFF, odd/even on
	IO_Write(0x3ce, 0x06);
	IO_Write(0x3cf, mono ? 0x0a : 0x0e);
```

johnnovak replied:
Cheers, yeah I split the lines first, then moved the comments around, that was my mistake 😄

johnnovak commented on src/dos/program_mode.cpp:
```diff
@@ -0,0 +1,293 @@
+	auto is_set_display_mode_command = [](const std::string& command) {
+		// Set display mode commands start with digits (e.g., 80x43)
+		constexpr auto MinLength = sizeof("80x43") - 1;
+		return isdigit(command[0]) && command.size() >= MinLength;
```
I'm doing that check in the function where I handle the video mode command.

japsmits commented on src/ints/int10_modes.cpp:
```diff
@@ -1949,51 +2056,67 @@ bool INT10_SetVideoMode(uint16_t mode)
 
-		IO_Write(crtc_base,0x38);IO_Write(crtc_base+1,0x48);	//Register lock 1
-		IO_Write(crtc_base,0x39);IO_Write(crtc_base+1,0xa5);	//Register lock 2
+		IO_Write(crtc_base,0x3a);
+		IO_Write(crtc_base+1,reg_3a);
+
+		IO_Write(crtc_base,0x31);
+		// Enable banked memory and 256k+ access
+		IO_Write(crtc_base+1,reg_31);
+
+		IO_Write(crtc_base,0x58);
+		// Enable 8 mb of linear addressing
+		IO_Write(crtc_base+1,0x3);
+
+		IO_Write(crtc_base,0x38);
+		// Register lock 1
+		IO_Write(crtc_base+1,0x48);
+
+		IO_Write(crtc_base,0x39);
+		// Register lock 2
+		IO_Write(crtc_base+1,0xa5);
```
I'll give you a hand. 😄 
```suggestion
		IO_Write(crtc_base,0x3a);
		IO_Write(crtc_base+1,reg_3a);

		// Enable banked memory and 256k+ access
		IO_Write(crtc_base,0x31);
		IO_Write(crtc_base+1,reg_31);

		// Enable 8 mb of linear addressing
		IO_Write(crtc_base,0x58);
		IO_Write(crtc_base+1,0x3);

		// Register lock 1
		IO_Write(crtc_base,0x38);
		IO_Write(crtc_base+1,0x48);

		// Register lock 2
		IO_Write(crtc_base,0x39);
		IO_Write(crtc_base+1,0xa5);
```

johnnovak replied:
Thanks @japsmits 😄

johnnovak commented:
Alright, @FeralChild64, I've changed things so I'm using the 132-column VESA modes now, so you got your 70 Hz cursor blink rate back 😎 

Plus a bunch of other cleanups, refactors, etc... the usual stuff.

@japsmits I believe I've addressed all your comments too.

FeralChild64 reviewed (approved):
I have no more remarks. Played a little with the new command, didn't find any problems now.

rderooy commented:
Would it make sense from a compatibility perspective with the MS-DOS mode command, to add support for the `co40`, `co80`, and `mono` mode options? Perhaps just as simple aliases for the 25x40 and 25x80 modes (assuming the emulated video adapter supports it)?

interloper98 commented:
Look at all those modes!

```c++
...
        { "40x25", 0x001}, { "80x25", 0x003}, { "80x28", 0x070},
        { "80x30", 0x071}, { "80x34", 0x072}, { "80x43", 0x073},
        { "80x50", 0x074}, { "80x60", 0x043}, {"132x25", 0x109},
        {"132x28", 0x230}, {"132x30", 0x231}, {"132x34", 0x232},
        {"132x43", 0x10a}, {"132x50", 0x10b}, {"132x60", 0x10c}
...
        { "40x25", 0x01}, { "80x25", 0x03}, { "80x28", 0x70},
        { "80x30", 0x71}, { "80x34", 0x72}, { "80x43", 0x73},
        { "80x50", 0x74}
...
```

Amazing, @johnnovak! Blows away the external tool I've been using.

![out](https://github.com/dosbox-staging/dosbox-staging/assets/165201780/39595524-1a27-4526-a76d-cb14ac29f7be)

johnnovak commented:
> Would it make sense from a compatibility perspective with the MS-DOS mode command, to add support for the `co40`, `co80`, and `mono` mode options? Perhaps just as simple aliases for the 25x40 and 25x80 modes (assuming the emulated video adapter supports it)?

@rderooy Yeah I can add the `co40` and `co80` aliases easily. Not sure adding a new `mono` mode is worth it, but I can alias it to 80x25 so at least the command succeeds.

I'll merge this then I'll do that in a separate PR.

johnnovak commented:
Thanks @FeralChild64 and @interloper98. Hopefully these new text modes won't leave people wanting for anything else 😄
